boa/src/builtins/uri/mod.rs

```diff
@@ -0,0 +1,168 @@
+use crate::builtins::BuiltIn;
+use crate::property::Attribute;
+use crate::{Context, JsValue, JsResult, JsString};
+use crate::object::function::make_builtin_fn;
+
+use percent_encoding::{utf8_percent_encode, AsciiSet, CONTROLS};
+use crate::builtins::string::code_point_at;
+
+/// https://url.spec.whatwg.org/#fragment-percent-encode-set
+const FRAGMENT: &AsciiSet = &CONTROLS.add(b' ').add(b'"').add(b'<').add(b'>').add(b'`');
+
+fn is_uri_mark(c: &str) -> bool {
+    return match c {
+        "-" => true,
+        "_" => true,
+        "." => true,
+        "!" => true,
+        "~" => true,
+        "*" => true,
+        "'" => true,
+        "(" => true,
+        ")" => true,
+        _ => false
+    }
+}
+
+fn is_uri_reserved(c: &str) -> bool {
+    return match c {
+        ";" => true,
+        "/" => true,
+        "?" => true,
+        ":" => true,
+        "@" => true,
+        "&" => true,
+        "=" => true,
+        "+" => true,
+        "$" => true,
+        "," => true,
+        _ => false
+    }
+}
+
+fn is_alpha_numeric(c: &str) -> bool {
+    let first_byte = c.as_bytes()[0];
+    return (97..=122).contains(&first_byte)
+        || (65..=90).contains(&first_byte)
+        || (48..=57).contains(&first_byte);
+}
+
+fn is_unescaped_uri_component_character(c: &str) -> bool {
+    return is_alpha_numeric(&c) || is_uri_mark(&c);
+}
+
+fn naive_decimal_to_hexadecimal(number: u32) -> String {
+    fn to_char(decimal_digit: usize) -> String {
+        let alpha = ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9", "A", "B", "C", "D", "E", "F"];
+        return alpha[decimal_digit].clone().to_string();
+    }
+
+    let mut remaining = number.clone();
+    let mut r = &number % 16;
+    let mut hex_number = "".to_string();
+
+    while &remaining - &r != 0 {
+        hex_number = to_char(r as usize) + &hex_number[..];
+        remaining = (remaining - r) / 16;
+        r = &remaining % 16;
+    }
+
+    return to_char(r as usize) + &hex_number[..];
+}
+
+#[derive(Debug, Clone, Copy)]
+pub(crate) struct Uri;
+
+impl BuiltIn for Uri {
```

Comment:
Implementing `BuiltIn` for `Uri` would create a new property with name `Uri` on the global object, making it not spec compliant.
My recommendation would be to skip the `BuiltIn` implementation for `Uri` and just create a new associated function `init`, then call it from the initialization function of the context without adding it to the array of global inits.